core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -1852,12 +1857,28 @@ allocate_iovec_app_buffer(wasm_module_inst_t module_inst,
     return __WASI_ESUCCESS;
 }
 
+static inline size_t
+min(size_t a, size_t b)
+{
+    return a > b ? b : a;
+}
+
 static wasi_errno_t
 copy_buffer_to_iovec_app(wasm_module_inst_t module_inst, uint8 *buf_begin,
-                         uint32 buf_size, iovec_app_t *data, uint32 data_len)
+                         uint32 buf_size, iovec_app_t *data, uint32 data_len,
+                         uint32 size_to_copy)
 {
     uint8 *buf = buf_begin;
     uint32 i;
+    uint32 size_to_copy_into_iovec;
+
+    if (buf_size < size_to_copy) {
+        return __WASI_EINVAL;
+    }
+
+    if (size_to_copy == 0) {
```

Comment:
wonder if we need this check given that we already have this check in L1893

Reply:
Yeah you're right. It saves one `validate_app_addr` call but then maybe it's preferred that if we pass in an invalid app addr, we always return `__WASI_EINVAL` even if the `size_to_copy` is 0